### PR TITLE
build lib/common/DataStructures

### PR DESCRIPTION
**Pull request by jianchun**
- Take more bits from coreclr pal.
- Disable a list of warnings (for now).
- Refactor/modify headers and add missing pieces.
- Ignore precompiled hearders so far.

lib/common/DataStructures builds.

### CODE REVIEW

**Comment by dilijev** [comment]
Don't worry about the failed copyright check for now. It's because the copyright script isn't in the linux branch (added here: d44c44401f474c4c6854a6134148b534ae3081b0)

**Comment by jianchun** [comment]
> Don't worry about the failed copyright check for now. It's because the copyright check script isn't in the linux branch (added here: d44c444).

@dilijev Thanks for the info! Yeah, all tests passed!

**Comment by dilijev** [comment]
@jianchun Are you adding an issue for the disabled warnings?

**Comment by dilijev** [comment]
@jianchun You could also cherry pick the pair of .py and .sh scripts that do the Copyright Check (but they have been changed a couple of times since they were introduced to improve them and fix bugs, so you'd need to cherry-pick those too). Might be best to merge the current state of master into the linux branch.

**Comment by ariya** [comment]
Exciting progress :+1:

**Comment by jianchun** [comment]
> Are you adding an issue for the disabled warnings?

@dilijev No. There would be too many. We are tracking with comment xplat-todo.

**Comment by jianchun** [comment]
> Exciting progress
@ariya Thanks!

File: lib/common/CommonBasic.h [hunk 1]
```diff
@@ -36,15 +44,134 @@ typedef wchar_t wchar16;
...
 
+#define _BitScanForward BitScanForward
+#define _BitScanForward64 BitScanForward64
+#define _BitScanReverse BitScanReverse
+#define _BitScanReverse64 BitScanReverse64
+#define _bittest BitTest
+#define _bittestandset BitTestAndSet
+#define _interlockedbittestandset InterlockedBitTestAndSet
+
+#ifdef PAL_STDCPP_COMPAT
```
Comment (curtisman) [inline lib/common/CommonBasic.h]: Can these be moved to a separate file?

File: lib/common/DataStructures/BaseDictionary.h [hunk 2]
```diff
@@ -34,6 +34,26 @@ namespace Js
 {
     template <class TDictionary>
     class RemoteDictionary;
+
+    class DefaultListLockPolicy
```
Comment (curtisman) [inline lib/common/DataStructures/BaseDictionary.h]: Separate file, and may be rename?  It doesn't just lock list, but dictionary too.

File: lib/common/Memory/HeapAllocator.h [hunk 3]
```diff
@@ -331,12 +331,8 @@ typedef NoThrowHeapAllocator NoThrowNoMemProtectHeapAllocator;
 #if !defined(USED_IN_STATIC_LIB)
 _Ret_maybenull_ void * __cdecl operator new(size_t byteSize);
 _Ret_maybenull_ void * __cdecl operator new[](size_t byteSize);
-
-// xplat-todo: Likely need to re-enable this at some point
-#ifdef _MSC_VER
-void __cdecl operator delete(void * obj);
-void __cdecl operator delete[](void * obj);
-#endif
+void __cdecl operator delete(void * obj) throw();
```
Comment (curtisman) [inline lib/common/Memory/HeapAllocator.h]: What the need for throw() here?

File: lib/common/CommonBasic.h (hunk 1 above)
Comment (jianchun) [inline lib/common/CommonBasic.h]: > Can these be moved to a separate file?

Sure, will do.

File: lib/common/Memory/HeapAllocator.h (hunk 3 above)
Comment (jianchun) [inline lib/common/Memory/HeapAllocator.h]: > Why the need for throw() here?

I got a clang warning that function was previously declared with an explicit exception specification. I'll revert the change here for now.

**Comment by dilijev** [comment]
@dotnet-bot test Copyright Check please

**Pull request merged**